Libraries/LibTTF/Font.cpp

```diff
@@ -0,0 +1,381 @@
+    }
+}
+
+i16 Hhea::ascender() const
+{
+    return be_i16(m_slice.offset_pointer((u32) Offsets::Ascender));
+}
+
+i16 Hhea::descender() const
+{
+    return be_i16(m_slice.offset_pointer((u32) Offsets::Descender));
+}
+
+i16 Hhea::line_gap() const
+{
+    return be_i16(m_slice.offset_pointer((u32) Offsets::LineGap));
+}
+
+u16 Hhea::advance_width_max() const
+{
+    return be_u16(m_slice.offset_pointer((u32) Offsets::AdvanceWidthMax));
+}
+
+u16 Hhea::number_of_h_metrics() const
+{
+    return be_u16(m_slice.offset_pointer((u32) Offsets::NumberOfHMetrics));
+}
+
+GlyphHorizontalMetrics Hmtx::get_glyph_horizontal_metrics(u32 glyph_id) const
+{
+    ASSERT(glyph_id < m_num_glyphs);
+    if (glyph_id < m_number_of_h_metrics) {
+        auto offset = glyph_id * (u32) Sizes::LongHorMetric;
+        u16 advance_width = be_u16(m_slice.offset_pointer(offset));
+        i16 left_side_bearing = be_i16(m_slice.offset_pointer(offset + 2));
+        return GlyphHorizontalMetrics {
+            .advance_width = advance_width,
+            .left_side_bearing = left_side_bearing,
+        };
+    }
+    auto offset = m_number_of_h_metrics * (u32) Sizes::LongHorMetric + (glyph_id - m_number_of_h_metrics) * (u32) Sizes::LeftSideBearing;
+    u16 advance_width = be_u16(m_slice.offset_pointer((m_number_of_h_metrics - 1) * (u32) Sizes::LongHorMetric));
+    i16 left_side_bearing = be_i16(m_slice.offset_pointer(offset));
+    return GlyphHorizontalMetrics {
+        .advance_width = advance_width,
+        .left_side_bearing = left_side_bearing,
+    };
+}
+
+u16 Maxp::num_glyphs() const
+{
+    return be_u16(m_slice.offset_pointer((u32) Offsets::NumGlyphs));
+}
+
+RefPtr<Font> Font::load_from_file(const StringView& path, unsigned index)
+{
+    auto file_or_error = Core::File::open(String(path), Core::IODevice::ReadOnly);
+    if (file_or_error.is_error()) {
+        dbg() << "Could not open file: " << file_or_error.error();
+        return nullptr;
+    }
+    auto file = file_or_error.value();
+    if (!file->open(Core::IODevice::ReadOnly)) {
+        dbg() << "Could not open file";
+        return nullptr;
+    }
+    auto buffer = file->read_all();
+    if (buffer.size() < 4) {
+        dbg() << "Font file too small";
+        return nullptr;
+    }
+    u32 tag = be_u32(buffer.data());
+    if (tag == tag_from_str("ttcf")) {
+        // It's a font collection
+        if (buffer.size() < (u32) Sizes::TTCHeaderV1 + sizeof(u32) * (index + 1)) {
+            dbg() << "Font file too small";
+            return nullptr;
+        }
+        u32 offset = be_u32(buffer.offset_pointer((u32) Sizes::TTCHeaderV1 + sizeof(u32) * index));
+        return adopt(*new Font(move(buffer), offset));
+    }
+    if (tag == tag_from_str("OTTO")) {
+        dbg() << "CFF fonts not supported yet";
+        return nullptr;
+    }
+    if (tag != 0x00010000) {
+        dbg() << "Not a valid  font";
+        return nullptr;
+    }
+    return adopt(*new Font(move(buffer), 0));
```

Comment:
It's more general than just this line,  but...
```suggestion
    return Font::load_from_offset(move(buffer), 0);
```

The idea is to move the header parsing code out of the constructor. By using a `static RefPtr<Font> load_from_offset(buffer, offset)` method instead, you can parse the header and split the table.
Then, once everything is parsed properly, you can create a font via `adopt(*new Font(move(head_slice), move(hhea_slice), ...)`.

The main advantage is gracious failures. Right now, if you try loading an invalid font file, you would just `ASSERT()` that the required tables are missing, which is not ideal (especially since this could happen in some important programs such as WindowServer, which does font rendering...).

It would also help with future-proofing, especially the load mechanism for WOFF/WOF2 formats (which requires to decompress some tables before being able to read them).